docs_src/design/adr/security/0026-authentication.md

```diff
@@ -0,0 +1,296 @@
+on the backend request that identify the caller,
+there is no mechanism by which Kong can prove to a
+backend service that it was the component that
+performed the authentication step.
+Even though the original JWT passes through the proxy,
+the Kong authentication plugins do not expose
+token introspection endpoints that the backend service
+could use to check token validity independently.
+
+The consequence of having an API gateway that performs
+all microservice authentication is that communication
+between EdgeX microservices running behind the API gateway
+are not authenticated in any way.
+EdgeX microservices are unable to distinguish
+malicous traffic that has evaded the API gateway
+from legitimate microservice traffic.
+
+
+## Decision
+
+This ADR expands the use of Vault to include
+issuing and validating [JWTs](https://www.rfc-editor.org/rfc/rfc7519)
+for use in microservice authentication among EdgeX services.
+Vault, and not Kong, owns JWTs issuance and validation
+in the new propsal.
+
+Having been relieved of JWT management responsibility,
+it is now possible to replace the Kong API gateway and its associated
+Postgres database with a much simpler (and smaller) NGINX reverse proxy.
+All inbound requests can be authorized by an external plugin
+that can in turn delegate the JWT validity check to Vault.
+From the perspective of an external caller,
+the existing JWT authentication feature is unchanged.
+TLS termination enabled by default so as to be consistent with
+[ADR 0015 - Encryption between microservices](./0015-in-cluster-tls.md).
+
+Behind the proxy, there are two major changes:
+
+1. Every EdgeX service, when security is enabled,
+   requires a JWT be passed as part of the HTTP request
+   that is validated using Vault's token introspection endpoint.
+
+2. Every EdgeX service, when security is enabled,
+   uses a Vault-supplied JWT to authenticate
+   outgoing calls to peer EdgeX services.
+   For microservice chaining scenarios,
+   the original caller's identity can be passed through
+   in most cases.
+
+The new TO-BE architecture is diagrammed in the following figure:
+
+![TO-BE Architecture](0026-to-be.jpg)
+
+The other architecturally signficant change to the proposal
+is the use Vault's identity features.
+Today, every EdgeX service is assigned a token with defined
+access to the EdgeX secret store.
+With this proposal, every service, and every external user,
+will will have an assigned Vault identity.
+The implication of this change is that it is now possible
+to configure Vault authentication engines against
+Vault identities.
+For example:
+
+* An external user identity could be authenticated by
+  an external service, such as [Auth0](https://auth0.com).
+  Alternatively, username/password or AppRole authentication
+  could be used if an external source of identity is not available.
+  This is viewed as beneficial, as downstream EdgeX implementations
+  are already building their own similar integrations.
```

Comment:
Would (or how would) a UI or tool be setup to create the user identity (or u/p or AppRole)?  Or would we care?

Reply:
Note to self: Edit above: Vault also has the ability to register a public key for JWT validation. Replace AppRole reference with JWT reference.

@jpwhitemn The `secrets-config` tool already supports creating an API gateway user.  We would just change it to create the user in Vault instead of Kong.   If someone integrates with an external identity management system, then that's out of scope for us.